controllers/awsmachinetemplate_controller.go

```diff
@@ -0,0 +1,393 @@
+	}
+
+	// Get region from AWSCluster (standard EC2-based cluster)
+	if cluster.Spec.InfrastructureRef != nil && cluster.Spec.InfrastructureRef.Kind == "AWSCluster" {
+		awsCluster := &infrav1.AWSCluster{}
+		if err := r.Get(ctx, client.ObjectKey{
+			Namespace: cluster.Namespace,
+			Name:      cluster.Spec.InfrastructureRef.Name,
+		}, awsCluster); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return "", errors.Wrapf(err, "failed to get AWSCluster %s/%s", cluster.Namespace, cluster.Spec.InfrastructureRef.Name)
+			}
+		} else if awsCluster.Spec.Region != "" {
+			return awsCluster.Spec.Region, nil
+		}
+	}
+
+	// Get region from AWSManagedControlPlane (EKS cluster)
+	if cluster.Spec.ControlPlaneRef != nil && cluster.Spec.ControlPlaneRef.Kind == "AWSManagedControlPlane" {
+		awsManagedCP := &ekscontrolplanev1.AWSManagedControlPlane{}
+		if err := r.Get(ctx, client.ObjectKey{
+			Namespace: cluster.Namespace,
+			Name:      cluster.Spec.ControlPlaneRef.Name,
+		}, awsManagedCP); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return "", errors.Wrapf(err, "failed to get AWSManagedControlPlane %s/%s", cluster.Namespace, cluster.Spec.ControlPlaneRef.Name)
+			}
+		} else if awsManagedCP.Spec.Region != "" {
+			return awsManagedCP.Spec.Region, nil
+		}
+	}
+
+	return "", nil
+}
+
+// getInstanceTypeCapacity queries AWS EC2 API for instance type capacity information.
+// Returns the resource list (CPU, Memory).
+func (r *AWSMachineTemplateReconciler) getInstanceTypeCapacity(ctx context.Context, ec2Client *ec2.Client, instanceType string) (corev1.ResourceList, error) {
+	// Query instance type information
+	input := &ec2.DescribeInstanceTypesInput{
+		InstanceTypes: []ec2types.InstanceType{ec2types.InstanceType(instanceType)},
+	}
+
+	result, err := ec2Client.DescribeInstanceTypes(ctx, input)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to describe instance type %q", instanceType)
+	}
+
+	if len(result.InstanceTypes) == 0 {
+		return nil, errors.Errorf("no information found for instance type %q", instanceType)
+	}
+
+	// Extract capacity information
+	info := result.InstanceTypes[0]
+	resourceList := corev1.ResourceList{}
+
+	// CPU
+	if info.VCpuInfo != nil && info.VCpuInfo.DefaultVCpus != nil {
+		resourceList[corev1.ResourceCPU] = *resource.NewQuantity(int64(*info.VCpuInfo.DefaultVCpus), resource.DecimalSI)
+	}
+
+	// Memory
+	if info.MemoryInfo != nil && info.MemoryInfo.SizeInMiB != nil {
+		resourceList[corev1.ResourceMemory] = resource.MustParse(fmt.Sprintf("%dMi", *info.MemoryInfo.SizeInMiB))
+	}
+
+	return resourceList, nil
+}
+
+// getNodeInfo queries node information (architecture and OS) for the AWSMachineTemplate.
+// It uses AMI ID if specified, otherwise attempts AMI lookup or falls back to instance type info.
+func (r *AWSMachineTemplateReconciler) getNodeInfo(ctx context.Context, ec2Client *ec2.Client, template *infrav1.AWSMachineTemplate, instanceType string) (*infrav1.NodeInfo, error) {
```

Comment:
This function is still a bit confusing to me. Can we have proper comments what we try to do?

And maybe use early returns instead of if elsing till the end.

E.g.:

```go
// Resolve via ami
if amiID := ptr.Deref(template.Spec.Template.Spec.AMI.ID, ""); amiID != "" {
  // do work
  return nodeInfo, nil
}

// Resolve via InstanceType
...
// do work
...
  return nodeInfo, nil
}

// Resolve via default AMI
...
return nodeinfo,nil

```